single_cell/annotation_import.R

```diff
@@ -0,0 +1,157 @@
+#' \item when columns containing notes exist, columns containing annotations must be distinguishable via a regex-definable naming scheme.
+#' The default expectation is that annotation column names start with "annot", while notes column names will not.
+#' }
+#' 
+#' How it works:\enumerate{
+#' \item If \code{object} is given an RDS file path rather than the Seurat or SCE object itself, the object is first loaded in
+#' \item The annotation csv is then read in with \code{read.csv(annots_file, header = TRUE, sep = sep)}
+#' \item The cell-metadata matching the first column name (clustering) of the annotations csv is extracted from the \code{object}
+#' \item An index map is built between values of this clustering metadata and values in this first column
+#' \item Columns matching the \code{annots_regex} regex are pulled in to the object making use of 1) the column name as the metadata name, 2) the pre-built index map.
+#' }
+#' 
+#' @author Daniel Bunis
+#' @importFrom dittoSeq getMetas
+#' @importFrom dittoSeq meta
+#' @examples
+#' # We'll use the Seurat example dataset for this example
+#' sobj <- SeuratObject::pbmc_small
+#' 
+#' # Our annots_file will be a csv version of:
+#' example_annotations <- data.frame(
+#'   RNA_snn_res.1 = 0:2,
+#'   annots_broad = c("T", "Myeloid", "B"),
+#'   notes = c("has CD3E expression", "has MS4A1 expression", "has CD14 expression")
+#' )
+#' print(example_annotations)
+#' 
+#' if (!file.exists("annotation_import_example.csv")) {
+#'   write.csv(
+#'     example_annotations,
+#'     file = "annotation_import_example.csv",
+#'     row.names = FALSE
+#'   )
+#' }
+#' 
+#' # By default, only columns whose name starts with "annot" will be pulled in
+#' sobj <- importAnnotations(
+#'   object = sobj,
+#'   annots_file = "annotation_import_example.csv")
+#'   
+#' # As we can see, the annotations are now pulled in:
+#' table(sobj$RNA_snn_res.1, sobj$annots_broad)
+#'
+#' # Adjust 'annots_regex' if needed, OR set it to NULL to pull in from all columns
+#' #   of the 'annots_file'.
+#' sobj <- importAnnotations(
+#'   object = sobj,
+#'   annots_file = "annotation_import_example.csv",
+#'   annots_regex = NULL)
+#'
+#' # Notice that in this second run, we are warned when the 'annots_broad' metadata
+#' #   which was already added in the first run, is being brought in again.
+#' # If you want to have the function error out instead of overwriting such
+#' #   metadata, set 'stop_if_overwrite' to TRUE.
+#' \dontrun{
+#' sobj <- importAnnotations(
+#'   object = sobj,
+#'   annots_file = "annotation_import_example.csv",
+#'   stop_if_overwrite = TRUE)
+#' }
+#' 
+importAnnotations <- function(
+    object,
+    annots_file,
+    annots_regex = "^annot",
+    sep = ",",
+    stop_if_overwrite = FALSE,
+    verbose = TRUE) {
+
+    # Until we decide a better way to handle this function I use all the time,
+    # simply adding this here, again
```

Comment:
can we put this in a central utils file that all essential scripts import?

Reply:
Not a bad idea.  I should make a separate PR for it, and then adjust my other scripts after... I think there's work in unifying Rebecca & Ravi & my versions that I'm sure have all diverged from Ravi's original one!